src/libraries/System.Private.DataContractSerialization/src/System/Runtime/Serialization/DataContract.cs

```diff
@@ -2060,42 +2030,113 @@ internal static void CheckAndAdd(Type type, Dictionary<Type, Type> typesChecked,
         {
             type = DataContract.UnwrapNullableType(type);
             DataContract dataContract = DataContract.GetDataContract(type);
-            DataContract? alreadyExistingContract;
             if (nameToDataContractTable == null)
             {
                 nameToDataContractTable = new DataContractDictionary();
             }
-            else if (nameToDataContractTable.TryGetValue(dataContract.StableName, out alreadyExistingContract))
+            else if (nameToDataContractTable.TryGetValue(dataContract.XmlName, out DataContract? alreadyExistingContract))
             {
+                // NOTE TODO smolloy - The existing contract type was used as-is in NetFx. The call to get the appropriate adapter type was added in CoreFx with https://github.com/dotnet/runtime/commit/50c0a70c52fa66fafa1227be552ccdab5e4cf8e4
                 // Don't throw duplicate if its a KeyValuePair<K,T> as it could have been added by Dictionary<K,T>
-                if (DataContractCriticalHelper.GetDataContractAdapterType(alreadyExistingContract.UnderlyingType) != DataContractCriticalHelper.GetDataContractAdapterType(type) &&
-                    !(alreadyExistingContract is ClassDataContract && ((ClassDataContract)alreadyExistingContract).IsKeyValuePairAdapter))
-                    throw System.Runtime.Serialization.DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.DupContractInKnownTypes, type, alreadyExistingContract.UnderlyingType, dataContract.StableName.Namespace, dataContract.StableName.Name)));
+                if (DataContractCriticalHelper.GetDataContractAdapterType(alreadyExistingContract.UnderlyingType) != DataContractCriticalHelper.GetDataContractAdapterType(type))
+                    throw System.Runtime.Serialization.DiagnosticUtility.ExceptionUtility.ThrowHelperError(new InvalidOperationException(SR.Format(SR.DupContractInKnownTypes, type, alreadyExistingContract.UnderlyingType, dataContract.XmlName.Namespace, dataContract.XmlName.Name)));
                 return;
             }
-            nameToDataContractTable.Add(dataContract.StableName, dataContract);
+            nameToDataContractTable.Add(dataContract.XmlName, dataContract);
             ImportKnownTypeAttributes(type, typesChecked, ref nameToDataContractTable);
         }
 
+        [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
+            Justification = "DataContract is an internal type, but the override here is the very public Object.Equals(). Being an internal type" +
+            "though, we control when this is called, and all callers are annotated with [RequiresUnreferencedCode].")]
```

Comment:
This justification is now incorrect. Do we need to do something different here?

Reply:
I have scaled back the propagation of [RUC] on this override. The only two places that actually need it stop the propagation chain with this same suppress attribute now, and are justified because they are instance methods on types whose consrtuctors all are marked with [RUC]. This justification is used in many places throughout the framework, such as in [System.Text.Json.Serialization.Converters.ObjectConverterFactory](https://source.dot.net/#System.Text.Json/System/Text/Json/Serialization/Converters/Object/ObjectConverterFactory.cs,36) for example.